modules/message-lane/src/lib.rs

```diff
@@ -90,19 +159,237 @@ decl_module! {
+		#[weight = DELIVERY_BASE_WEIGHT + max_dispatch_weight]
+		pub fn receive_messages_proof(origin, proof: MessagesProofOf<T, I>, max_dispatch_weight: Weight) -> DispatchResult {
+			let relayer = ensure_signed(origin)?;
+
+			// verify messages proof && convert proof into messages
+			let messages = T::SourceHeaderChain::verify_messages_proof(proof).map_err(|err| {
+				frame_support::debug::trace!(
+					target: "runtime",
+					"Rejecting invalid messages proof: {:?}",
+					err,
+				);
+
+				Error::<T, I>::InvalidMessagesProof
+			})?;
+
+			// estimate cumulative weight of messages dispatch
+			let mut dispatcher = T::MessageDispatch::with_allowed_weight(max_dispatch_weight);
+			let messages_dispatch_weight: Weight = messages.iter().map(|m| dispatcher.dispatch_weight(m)).sum();
+
+			// the submitter pays for dispatch in advance, so the whole `messages_dispatch_weight` should
+			// be covered. But it has already paid regular fee for at least `max_dispatch_weight`, so we
+			// only need to withdraw `messages_dispatch_weight - max_dispatch_weight`.
+			let messages_dispatch_extra_weight = messages_dispatch_weight.saturating_sub(max_dispatch_weight);
+			T::MessageDispatchPayment::pay_dispatch_fee(&relayer, messages_dispatch_extra_weight).map_err(|err| {
+				frame_support::debug::trace!(
+					target: "runtime",
+					"Rejecting {} messages because relayer {:?} is unable to pay extra fee for weight {}: {:?}",
+					messages.len(),
+					relayer,
+					messages_dispatch_weight,
+					err,
+				);
+
+				Error::<T, I>::FailedToWithdrawDispatchFee
+			})?;
+
+			// now let's process as many messages, as possible
+			let total_messages = messages.len();
+			let mut invalid_messages = 0;
+			let mut processed_messages = 0;
+			let mut queued_messages = 0;
+			for message in messages {
+				let mut lane = inbound_lane::<T, I>(message.key.lane_id);
+				match lane.receive_message(message.key.nonce, message.data, &mut dispatcher) {
+					ReceiveMessageResult::Invalid => invalid_messages += 1,
+					ReceiveMessageResult::Processed => processed_messages += 1,
+					ReceiveMessageResult::Queued => queued_messages += 1,
+				}
+			}
+
+			let weight_left = dispatcher.weight_left();
+			frame_support::debug::trace!(
+				target: "runtime",
+				"Received messages: total={}, invalid={}, processed={}, queued={}. Weight spent: {}, left: {}",
+				total_messages,
+				invalid_messages,
+				processed_messages,
+				queued_messages,
+				max_dispatch_weight.saturating_sub(weight_left),
+				weight_left,
+			);
+
+			// ok - we have processed or queued all bundled messages. But what if relayer acts aggressivly
+			// (i.e. `max_dispatch_weight` is larger than `messages_dispatch_weight`), or if some messages
+			// have been queued? Then we may have some capacity left from `max_dispatch_weight`.
+			// => let's not waste it and try to process queued messages
+			let mut queued_dispatcher = T::QueuedMessageDispatch::with_allowed_weight(weight_left);
```

Comment:
That's cool, but it means that we actually pay twice for delivering/dispatching these messages. I'd say that here, the relayer who is willing to spend more `max_dispatch_weight` to process extra messages from the queue, should be rewarded for that, by getting back the fee (or some portion of it) payed by the less aggressive relayer, who queued these messages. wdyt?

Currently, the strategy for relayers is to send smallest possible `max_dispatch_weight`, cause they pay small amount upfront and they don't overpay in case the transaction fails (imagine there is a race between multiple relayers). Restricting the minimal value of `max_dispatch_weight` + giving them some incentive to actually process messages in the queue might help with overall bridge efficiency.

Reply:
> That's cool, but it means that we actually pay twice for delivering/dispatching these messages

Why? Who's paying twice? Relayer is only paying base delivery tx fee + dispatch fee.

>  I'd say that here, the relayer who is willing to spend more max_dispatch_weight to process extra messages from the queue, should be rewarded for that, by getting back the fee (or some portion of it) payed by the less aggressive relayer, who queued these messages. wdyt?

If there's relayer that is submitting transactions with `max_dispatch_weight` larger than `weights-sum(delivered-messages)`, then either (1) it is alternatively-incentivised to dispatch messages, not to deliver messages - that should be controlled on app-level imo (2) it is trying to fill block with dummy transactions (possibly malicious or dumb) - I don't think we should reward this behavior.

Remember that reserving more weight than necessary decreases transactions throughput. It only works if relayers are paying for such transactions. Rewarding this behavior may result in blocks spammed with dummy transactions.

Reply:
> Why? Who's paying twice?

The first relayer to deliver these queued messages is already charged with `fee` for this delivery&dispatch. Now here the relayer is paying the `fee` again, since it declared high `max_dispatch_weight` to actually process these queued messages, so it's charged a `fee` for that declared weight. Unless I'm missing something.

> it is trying to fill block with dummy transactions (possibly malicious or dumb) - I don't think we should reward this behavior.

Processing messages in the queue is not really malicious. If they pay higher fee than other `target` blockchain users, it's totally legit for them to spend it even just idle-looping, if they actually make progress on the bridge lane, then it's better for us.

Also notice, that I'm not mentioning that we should mint coins out of thing air to reward them, I suggested that the relayer, how decided to queue the messages instead of processing them (because they thought it would be easier for them to get included in a block with lower-weight transaction) pays that extra price. I.e. it then creates a market for relayers between: delivering & queuing (low weight, higher fee, lower reward) or delivering & processing (high weight, lower fee, higher reward), so the relayers are now responsible for finding an optimum balance between this two sides and given current blockchain dynamics (other transactions waiting in the `target` blockchain transaction pool).

Reply:
Irrelevant